windows/src/developer/TIKE/compile/MergeKeyboardInfo.pas

```diff
@@ -418,6 +424,114 @@ procedure TMergeKeyboardInfo.CheckOrAddID;
     json.AddPair('id', FID);
 end;
 
+procedure TMergeKeyboardInfo.AddSubtagNames(id: String; var o: TJSONObject);
+var
+  displayName, languageName, scriptName, regionName: String;
+  v: TJSONValue;
+begin
+  if id = '' then
+    Exit;
+  with TBCP47Tag.Create(id) do
+  try
+    TLanguageCodeUtils.BCP47Languages.TryGetValue(Language, languageName);
+    TLanguageCodeUtils.BCP47Scripts.TryGetValue(Script, scriptName);
+    TLanguageCodeUtils.BCP47Regions.TryGetValue(Region, regionName);
+  finally
+    Free;
+  end;
+
+  displayName := TLanguageCodeUtils.LanguageName(languageName, scriptName, regionName);
+
+  v := o.Values[TKeyboardInfoFile.SDisplayName];
+  if not Assigned(v) then
+    o.AddPair(TKeyboardInfoFile.SDisplayName, displayName);
+
+  v := o.Values[TKeyboardInfoFile.SLanguageName];
+  if not Assigned(v) then
+    o.AddPair(TKeyboardInfoFile.SLanguageName, languageName);
+
+  v := o.Values[TKeyboardInfoFile.SScriptName];
+  if not Assigned(v) and (scriptName <> '') then
+    o.AddPair(TKeyboardInfoFile.SScriptName, scriptName);
+
+  v := o.Values[TKeyboardInfoFile.SRegionName];
+  if not Assigned(v) and (regionName <> '') then
+    o.AddPair(TKeyboardInfoFile.SRegionName, regionName);
+end;
+
+function TMergeKeyboardInfo.CheckOrMigrateLanguages: Boolean;
+var
+  v: TJSONValue;
+  alangs: TJSONArray;
+  olangs: TJSONObject; 
+  o: TJSONObject;
+  i: Integer;
+  id: string;
+begin
+  Result := True;
+  
+  v := json.GetValue(TKeyboardInfoFile.SLanguages);
+
+  // Migrate languages[] array to Object
+  if v is TJSONArray then
+  begin
+    alangs := v as TJSONArray;
+    olangs := TJSONObject.Create;
+    try
+      for i := 0 to alangs.Count - 1 do
+      begin
+        id := alangs.Items[i].Value;
+        if id = '' then
+          continue;
+
+        o := TJSONObject.Create;
+        try
+          olangs.AddPair(id, o);
+          AddSubtagNames(id, o);
+        finally
+          o.Free;
+        end;
+      end;
+
+      json.RemovePair(TKeyboardInfoFile.SLanguages);
+      json.AddPair(TKeyboardInfoFile.SLanguages, olangs);
+
+      // Migration complete and subtags populated
+      Exit;
```

Comment:
This shouldn't be needed -- also clearer to do it outside a `try`/`finally` block if you do need it.